### ISSUE TITLE
Enable dma_bench app module

### ISSUE THREAD

**Issue by multi-stream** [corundum/corundum#128]
How do I enable app module (dma_bench specifically) and integrate into the mqnic design?

**Comment by alexforencich** [comment]
Make a copy of the makefile and config.tcl for one of the design variants for your board, then make the appropriate changes.

Here is an example: https://github.com/corundum/corundum/tree/master/fpga/mqnic/fb2CG/fpga_100g/fpga_app_dma_bench

Basically, all you need to do is set the application block configuration in config.tcl and add all of the required files to the makefile.  So, for example, dma_bench requires config.tcl to be set like so:

    # Application block configuration
    dict set params APP_ID "32'h12348001"
    dict set params APP_ENABLE "1"
    dict set params APP_CTRL_ENABLE "1"
    dict set params APP_DMA_ENABLE "1"
    dict set params APP_AXIS_DIRECT_ENABLE "1"
    dict set params APP_AXIS_SYNC_ENABLE "1"
    dict set params APP_AXIS_IF_ENABLE "1"
    dict set params APP_STAT_ENABLE "1"

The application module is responsible for validating the settings (i.e. it should check APP_ID and the various _ENABLE parameters, and fail the build if the configuration is not acceptable).  For example, the DMA benchmark application really only requires APP_DMA_ENABLE, all of the others can be turned off as all of those interfaces are either tied off or looped through as appropriate.

And then the makefile needs to have

    SYN_FILES += app/dma_bench/rtl/mqnic_app_block_dma_bench.v

This one is simple and doesn't require any additional modules beyond the main application module, but in principle other application modules might require more files.

**Comment by multi-stream** [comment]
Thanks for help Alex! This builds the overall design correctly for AU50. However, the mqnic_app_dma_bench.ko is crashing the kernel. The NIC function also stopped working. Is this expected? Thanks!

[28427.900941] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: mqnic_app_dma_bench_probe() called
[28427.900966] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: Allocated DMA region virt 00000000348324b4, phys 00000000652cff22
[28427.900972] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: Statistics counters
[28427.900977] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_rx_tlp_mem_rd: 0
[28427.900982] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_rx_tlp_mem_wr: 0
[28427.900986] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_rx_tlp_io: 0
[28427.900990] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_rx_tlp_cfg: 0
[28427.900995] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_rx_tlp_msg: 0
[28427.900998] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_rx_tlp_cpl: 0
[28427.901002] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_rx_tlp_cpl_ur: 0
[28427.901006] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_rx_tlp_cpl_ca: 0
[28427.901010] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_rx_tlp_atomic: 0
[28427.901014] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_rx_tlp_ep: 0
[28427.901018] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_rx_tlp_hdr_dw: 0
[28427.901021] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_rx_tlp_req_dw: 0
[28427.901025] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_rx_tlp_payload_dw: 0
[28427.901029] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_rx_tlp_cpl_dw: 0
[28427.901033] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_tx_tlp_mem_rd: 0
[28427.901037] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_tx_tlp_mem_wr: 0
[28427.901041] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_tx_tlp_io: 0
[28427.901044] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_tx_tlp_cfg: 0
[28427.901048] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_tx_tlp_msg: 0
[28427.901052] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_tx_tlp_cpl: 0
[28427.901056] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_tx_tlp_cpl_ur: 0
[28427.901059] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_tx_tlp_cpl_ca: 0
[28427.901063] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_tx_tlp_atomic: 0
[28427.901067] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_tx_tlp_ep: 0
[28427.901071] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_tx_tlp_hdr_dw: 0
[28427.901074] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_tx_tlp_req_dw: 0
[28427.901078] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_tx_tlp_payload_dw: 0
[28427.901082] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_tx_tlp_cpl_dw: 0
[28427.901086] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_rd_op_count: 0
[28427.901089] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_rd_op_bytes: 0
[28427.901093] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_rd_op_latency: 0
[28427.901097] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_rd_op_error: 0
[28427.901101] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_rd_req_count: 0
[28427.901104] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_rd_req_latency: 0
[28427.901108] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_rd_req_timeout: 0
[28427.901112] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_rd_op_table_full: 0
[28427.901116] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_rd_no_tags: 0
[28427.901120] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_rd_tx_no_credit: 0
[28427.901124] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_rd_tx_limit: 0
[28427.901128] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_rd_tx_stall: 0
[28427.901132] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_wr_op_count: 0
[28427.901136] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_wr_op_bytes: 0
[28427.901140] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_wr_op_latency: 0
[28427.901144] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_wr_op_error: 0
[28427.901147] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_wr_req_count: 0
[28427.901151] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_wr_req_latency: 0
[28427.901155] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_wr_op_table_full: 0
[28427.901159] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_wr_tx_no_credit: 0
[28427.901163] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_wr_tx_limit: 0
[28427.901167] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_wr_tx_stall: 0
[28427.901169] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: Write test data
[28427.901172] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: Read test data
[28427.901174] 00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f  ................
[28427.901177] 10 11 12 13 14 15 16 17 18 19 1a 1b 1c 1d 1e 1f  ................
[28427.901179] 20 21 22 23 24 25 26 27 28 29 2a 2b 2c 2d 2e 2f   !"#$%&'()*+,-./
[28427.901180] 30 31 32 33 34 35 36 37 38 39 3a 3b 3c 3d 3e 3f  0123456789:;<=>?
[28427.901182] 40 41 42 43 44 45 46 47 48 49 4a 4b 4c 4d 4e 4f  @ABCDEFGHIJKLMNO
[28427.901184] 50 51 52 53 54 55 56 57 58 59 5a 5b 5c 5d 5e 5f  PQRSTUVWXYZ[\]^_
[28427.901185] 60 61 62 63 64 65 66 67 68 69 6a 6b 6c 6d 6e 6f  `abcdefghijklmno
[28427.901187] 70 71 72 73 74 75 76 77 78 79 7a 7b 7c 7d 7e 7f  pqrstuvwxyz{|}~.
[28427.901189] 80 81 82 83 84 85 86 87 88 89 8a 8b 8c 8d 8e 8f  ................
[28427.901191] 90 91 92 93 94 95 96 97 98 99 9a 9b 9c 9d 9e 9f  ................
[28427.901192] a0 a1 a2 a3 a4 a5 a6 a7 a8 a9 aa ab ac ad ae af  ................
[28427.901194] b0 b1 b2 b3 b4 b5 b6 b7 b8 b9 ba bb bc bd be bf  ................
[28427.901196] c0 c1 c2 c3 c4 c5 c6 c7 c8 c9 ca cb cc cd ce cf  ................
[28427.901197] d0 d1 d2 d3 d4 d5 d6 d7 d8 d9 da db dc dd de df  ................
[28427.901199] e0 e1 e2 e3 e4 e5 e6 e7 e8 e9 ea eb ec ed ee ef  ................
[28427.901201] f0 f1 f2 f3 f4 f5 f6 f7 f8 f9 fa fb fc fd fe ff  ................
[28427.901203] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: Start copy to card
[28427.901211] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: Start copy to host
[28427.901217] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: read test data
[28427.901219] 00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f  ................
[28427.901221] 10 11 12 13 14 15 16 17 18 19 1a 1b 1c 1d 1e 1f  ................
[28427.901223] 20 21 22 23 24 25 26 27 28 29 2a 2b 2c 2d 2e 2f   !"#$%&'()*+,-./
[28427.901224] 30 31 32 33 34 35 36 37 38 39 3a 3b 3c 3d 3e 3f  0123456789:;<=>?
[28427.901226] 40 41 42 43 44 45 46 47 48 49 4a 4b 4c 4d 4e 4f  @ABCDEFGHIJKLMNO
[28427.901228] 50 51 52 53 54 55 56 57 58 59 5a 5b 5c 5d 5e 5f  PQRSTUVWXYZ[\]^_
[28427.901230] 60 61 62 63 64 65 66 67 68 69 6a 6b 6c 6d 6e 6f  `abcdefghijklmno
[28427.901231] 70 71 72 73 74 75 76 77 78 79 7a 7b 7c 7d 7e 7f  pqrstuvwxyz{|}~.
[28427.901233] 80 81 82 83 84 85 86 87 88 89 8a 8b 8c 8d 8e 8f  ................
[28427.901235] 90 91 92 93 94 95 96 97 98 99 9a 9b 9c 9d 9e 9f  ................
[28427.901236] a0 a1 a2 a3 a4 a5 a6 a7 a8 a9 aa ab ac ad ae af  ................
[28427.901238] b0 b1 b2 b3 b4 b5 b6 b7 b8 b9 ba bb bc bd be bf  ................
[28427.901240] c0 c1 c2 c3 c4 c5 c6 c7 c8 c9 ca cb cc cd ce cf  ................
[28427.901242] d0 d1 d2 d3 d4 d5 d6 d7 d8 d9 da db dc dd de df  ................
[28427.901244] e0 e1 e2 e3 e4 e5 e6 e7 e8 e9 ea eb ec ed ee ef  ................
[28427.901246] f0 f1 f2 f3 f4 f5 f6 f7 f8 f9 fa fb fc fd fe ff  ................
[28427.901248] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: test data matches
[28427.901250] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: perform block reads (dma_alloc_coherent)
[28427.903597] divide error: 0000 [#1] PREEMPT SMP NOPTI
[28427.903635] CPU: 12 PID: 62474 Comm: insmod Tainted: G        W IOE     5.19.0-rc1+ #7
[28427.903671] Hardware name: Dell Inc. PowerEdge R740xd/0YNX56, BIOS 2.15.1 06/15/2022
[28427.903701] RIP: 0010:dma_block_read_bench.constprop.0+0x303/0x336 [mqnic_app_dma_bench]
[28427.903738] Code: 69 c5 00 b4 c4 04 4e 8d 0c a5 00 00 00 00 4c 2b 45 c0 48 c7 c6 68 46 c9 c0 49 f7 f1 31 d2 50 4a 8d 04 85 00 00 00 00 4d 89 f0 <49> f7 f7 31 d2 50 48 8b 45 d0 41 57 48 c1 e0 02 48 f7 75 b8 ba 10
[28427.903803] RSP: 0018:ffffa3b025b1f960 EFLAGS: 00010246
[28427.903830] RAX: 0000000000000000 RBX: ffff8ed39e401228 RCX: 0000000000000001
[28427.903860] RDX: 0000000000000000 RSI: ffffffffc0c94668 RDI: ffff8ed39ddef800
[28427.903890] RBP: ffffa3b025b1f9b0 R08: 0000000000000001 R09: 0000000000232208
[28427.903917] R10: ffffffffffffffff R11: 0000000000000067 R12: 000000000008c882
[28427.903945] R13: 0000000000000001 R14: 0000000000000001 R15: 0000000000000000
[28427.903974] FS:  00007f6c08151000(0000) GS:ffff8ed2c0f80000(0000) knlGS:0000000000000000
[28427.904006] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[28427.904031] CR2: 0000564e70115c50 CR3: 00000001143a6001 CR4: 00000000007706e0
[28427.904059] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[28427.904087] DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
[28427.904116] PKRU: 55555554
[28427.904130] Call Trace:
[28427.904145]  <TASK>
[28427.904161]  mqnic_app_dma_bench_probe+0x44b/0x693 [mqnic_app_dma_bench]
[28427.904194]  ? acpi_dev_pm_attach+0x25/0x110
[28427.904223]  ? dma_block_read_bench.constprop.0+0x336/0x336 [mqnic_app_dma_bench]
[28427.904256]  auxiliary_bus_probe+0x47/0x80
[28427.904281]  ? driver_sysfs_add+0x77/0xe0
[28427.904305]  really_probe+0x1d0/0x3b0
[28427.904328]  __driver_probe_device+0x116/0x190
[28427.904352]  driver_probe_device+0x23/0xc0
[28427.904375]  __driver_attach+0xbd/0x1e0
[28427.904398]  ? __device_attach_driver+0x110/0x110
[28427.904423]  bus_for_each_dev+0x7c/0xc0
[28427.904446]  driver_attach+0x1e/0x30
[28427.904468]  bus_add_driver+0x171/0x210
[28427.904490]  driver_register+0x95/0xf0
[28427.904513]  __auxiliary_driver_register+0x6f/0xd0
[28427.904537]  ? 0xffffffffc0cbd000
[28427.904556]  mqnic_app_dma_bench_init+0x23/0x1000 [mqnic_app_dma_bench]
[28427.904587]  do_one_initcall+0x46/0x210
[28427.904613]  ? kmem_cache_alloc_trace+0x1a6/0x320
[28427.904642]  do_init_module+0x52/0x210
[28427.904666]  load_module+0x2012/0x2420
[28427.904693]  __do_sys_finit_module+0xc5/0x130
[28427.904716]  ? __do_sys_finit_module+0xc5/0x130
[28427.904744]  __x64_sys_finit_module+0x18/0x20
[28427.904767]  do_syscall_64+0x59/0x80
[28427.905521]  ? ksys_mmap_pgoff+0x10d/0x250
[28427.906201]  ? exit_to_user_mode_prepare+0x35/0x170
[28427.906880]  ? syscall_exit_to_user_mode+0x26/0x50
[28427.907545]  ? __x64_sys_mmap+0x33/0x40
[28427.908195]  ? do_syscall_64+0x69/0x80
[28427.908832]  entry_SYSCALL_64_after_hwframe+0x46/0xb0
[28427.909465] RIP: 0033:0x7f6c0791ea3d
[28427.910073] Code: 5b 41 5c c3 66 0f 1f 84 00 00 00 00 00 f3 0f 1e fa 48 89 f8 48 89 f7 48 89 d6 48 89 ca 4d 89 c2 4d 89 c8 4c 8b 4c 24 08 0f 05 <48> 3d 01 f0 ff ff 73 01 c3 48 8b 0d c3 a3 0f 00 f7 d8 64 89 01 48
[28427.911317] RSP: 002b:00007fffcbaf2d18 EFLAGS: 00000246 ORIG_RAX: 0000000000000139
[28427.911885] RAX: ffffffffffffffda RBX: 0000564e715a27e0 RCX: 00007f6c0791ea3d
[28427.912147] RDX: 0000000000000000 RSI: 0000564e7011ecd2 RDI: 0000000000000003
[28427.912412] RBP: 0000000000000000 R08: 0000000000000000 R09: 0000000000000000
[28427.912676] R10: 0000000000000003 R11: 0000000000000246 R12: 0000564e7011ecd2
[28427.912937] R13: 0000564e715a2760 R14: 0000564e7011d888 R15: 0000564e715a2900
[28427.913201]  </TASK>
[28427.913453] Modules linked in: mqnic_app_dma_bench(OE+) tls mqnic(OE) nfsv3 nfs_acl rpcsec_gss_krb5 nfsv4 nfs lockd grace fscache netfs binfmt_misc nvme_fabrics intel_rapl_msr intel_rapl_common intel_uncore_frequency intel_uncore_frequency_common isst_if_common skx_edac nfit x86_pkg_temp_thermal intel_powerclamp coretemp kvm_intel ipmi_ssif kvm cmdlinepart spi_nor rapl dell_smbios mei_me dcdbas intel_cstate joydev dell_wmi_descriptor wmi_bmof input_leds mei mtd intel_pch_thermal acpi_ipmi ipmi_si acpi_power_meter mac_hid auth_rpcgss sch_fq_codel sunrpc ipmi_devintf dm_multipath ipmi_msghandler scsi_dh_rdac scsi_dh_emc scsi_dh_alua ramoops msr pstore_blk reed_solomon pstore_zone efi_pstore ip_tables x_tables autofs4 btrfs blake2b_generic zstd_compress raid10 raid456 async_raid6_recov async_memcpy async_pq async_xor async_tx xor raid6_pq libcrc32c raid1 raid0 multipath linear mlx4_ib ib_uverbs ib_core mlx4_en hid_generic usbhid hid mgag200 drm_shmem_helper drm_kms_helper syscopyarea
[28427.913510]  sysfillrect crct10dif_pclmul crc32_pclmul ghash_clmulni_intel sysimgblt nvme fb_sys_fops ixgbe aesni_intel igb xfrm_algo crypto_simd spi_intel_pci bnxt_en mlx4_core i2c_i801 drm cryptd nvme_core megaraid_sas ahci xhci_pci dca spi_intel lpc_ich i2c_smbus libahci xhci_pci_renesas wmi i2c_algo_bit mdio
[28427.916972] ---[ end trace 0000000000000000 ]---
[28427.954067] RIP: 0010:dma_block_read_bench.constprop.0+0x303/0x336 [mqnic_app_dma_bench]
[28427.954456] Code: 69 c5 00 b4 c4 04 4e 8d 0c a5 00 00 00 00 4c 2b 45 c0 48 c7 c6 68 46 c9 c0 49 f7 f1 31 d2 50 4a 8d 04 85 00 00 00 00 4d 89 f0 <49> f7 f7 31 d2 50 48 8b 45 d0 41 57 48 c1 e0 02 48 f7 75 b8 ba 10
[28427.954971] RSP: 0018:ffffa3b025b1f960 EFLAGS: 00010246
[28427.955236] RAX: 0000000000000000 RBX: ffff8ed39e401228 RCX: 0000000000000001
[28427.955502] RDX: 0000000000000000 RSI: ffffffffc0c94668 RDI: ffff8ed39ddef800
[28427.955784] RBP: ffffa3b025b1f9b0 R08: 0000000000000001 R09: 0000000000232208
[28427.956053] R10: ffffffffffffffff R11: 0000000000000067 R12: 000000000008c882
[28427.956322] R13: 0000000000000001 R14: 0000000000000001 R15: 0000000000000000
[28427.956593] FS:  00007f6c08151000(0000) GS:ffff8ed2c0f80000(0000) knlGS:0000000000000000
[28427.956868] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[28427.957143] CR2: 0000564e70115c50 CR3: 00000001143a6001 CR4: 00000000007706e0
[28427.957419] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[28427.957695] DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
[28427.957982] PKRU: 55555554

**Comment by alexforencich** [comment]
Ah I see, it crashed with a divide by zero error.  Looks like it's reading the statistics counters as all zeros, so there is probably an issue with how that module is reading the counters.  I'll take a look and see if I can figure out what's going on.

**Comment by alexforencich** [comment]
Try with the code from this repo: https://github.com/ucsdsysnet/corundum .  This should be fixed to read the counters correctly.

**Comment by multi-stream** [comment]
Thanks Alex! This works well now. And just one more thing: The dma_bench app breaks the nic function. I assume dma_bench is not intended for nic function to work at the same time, right? Thanks!

**Comment by alexforencich** [comment]
Define "breaks the NIC function".  Does it not work at all?  Does it stop working after loading the extension module?  Is something else going on?  The application logic should be completely independent of the core datapath, so aside from sharing the BW everything should work normally.  

I did notice that when I loaded the extension module during my testing last week, the write DMA BW was not what it should be (~75 Gbps, instead of ~100 Gbps), so perhaps there is a bug somewhere that I need to take a closer look at.

**Comment by multi-stream** [comment]
So ping to peer server simply stops working. Looking closer within the driver on the tx path, it looks like the packet is created but tx dma is not proceeding when the tx_queue head pointer is updated.

**Comment by alexforencich** [comment]
Okay, does this always happen with this design, or for example only after loading the extension kernel module?

**Comment by multi-stream** [comment]
This always happens with or without the extension kernel module.

**Comment by alexforencich** [comment]
Strange.  I will take a closer look, that shouldn't be the case.

**Comment by alexforencich** [comment]
I made some headway on this; I was able to replicate the issue in HW, and it seems that for some reason the card is not issuing any interrupts.  The event queues have data when they should be empty, and /proc/interrupts shows no activity.  Are you seeing the same on your end?  I have not yet been able to replicate the issue in simulation, however.

**Comment by alexforencich** [comment]
Ah.  Looks like something gets screwed up in the PCIe core config for MSI-X when the application section is enabled.

Working config, as reported by lspci:

	Capabilities: [60] MSI-X: Enable- Count=64 Masked-
		Vector table: BAR=0 offset=00010000
		PBA: BAR=0 offset=00018000

Broken config, as reported by lspci:

	Capabilities: [60] MSI-X: Enable- Count=64 Masked-
		Vector table: BAR=0 offset=00000040
		PBA: BAR=0 offset=00000050

So, the OS is looking for the MSI-X table in the wrong spot and things don't get configured correctly.  

Apparently when the application BAR is enabled, the MSI-X settings get changed and need to be re-applied.  Frapping vivado!  This should just be a change to some TCL scripts, I will let you know when I get it sorted out.  

Also, I refactored the DMA bench application and peeled the actual DMA benchmark portion into a separate verilog module.  It's looking like the DMA bench application could be a good vehicle for testing various aspects of the design, particularly the memory interfaces, so I went ahead and added targets for all of the Alveo boards, and will eventually add it for all of the boards.

**Comment by alexforencich** [comment]
Try the latest code at https://github.com/ucsdsysnet/corundum .  This at least fixes the MSI-X problem so the main NIC functionality will work.  However, the DMA write test is still reporting ~76 Gbps.  Interestingly, the NIC RX still hits 100G line rate, so it looks like I still have some debugging to do.

**Comment by alexforencich** [comment]
Fixed the DMA write BW limit, but I'm still not sure exactly what was going on.  Basically, PCIe flow control used to be handled more directly by the DMA engines, and now it's handled in the shims.  Initially, the number of operations in the TX pipeline was limited so that backpressure from the PCIe IP core would work correctly.  But then I rewrote the shims to take advantage of the segmented interface, and in doing so I also had to move the flow control into the shim since the shim now contained some largeish FIFOs.  When I did that, I didn't change the number of in-flight writes, so it was left at an unreasonably low number (3).  Changing that to a significantly larger number has improved the write DMA throughput, but what I don't understand at this point is why the NIC was able to issue DMA writes faster than the DMA benchmarking code, since the limit applied to the DMA engine itself and both components use the exact same DMA engine.

**Comment by alexforencich** [comment]
Should be fully fixed now at https://github.com/ucsdsysnet/corundum.  I will be getting that merged back into this repo relatively soon.

**Comment by alexforencich** [comment]
And the main repo has been updated.  Please let me know if you're running in to any other issues with the DMA benchmark design.

**Comment by multi-stream** [comment]
Hi Alex, Thanks for taking the time investigate the issue. 
It's been busy at the office. I'll try building it from the main repo tomorrow morning and report back here as soon as I can.
Thanks!

**Comment by cmegias** [comment]
I have tested it on Nexus K3P-S card and the NIC function remains working. And the output shown when running dmesg after loading the extension module seems to be fine.

**Comment by multi-stream** [comment]
HI Alex! This works with the AU50 on a Intel Xeon machine. However, when I moved the card to an ARM based platform the tx dma is not working. I can see the ping from peer and ARP reply packet is created, but the tx dma seems to be stuck and no follow up tx dma completion interrupt. Could you shed some light where we should look? 
BTW, this goes beyond my original question. I can open a new issue if that works better. Thanks!

**Comment by lizajoseph** [comment]
> Make a copy of the makefile and config.tcl for one of the design variants for your board, then make the appropriate changes.
> 
> Here is an example: https://github.com/corundum/corundum/tree/master/fpga/mqnic/fb2CG/fpga_100g/fpga_app_dma_bench
> 
> Basically, all you need to do is set the application block configuration in config.tcl and add all of the required files to the makefile. So, for example, dma_bench requires config.tcl to be set like so:
> 
> 
# Application block configuration
> dict set params APP_ID "32'h12348001"
> dict set params APP_ENABLE "1"
> dict set params APP_CTRL_ENABLE "1"
> dict set params APP_DMA_ENABLE "1"
> dict set params APP_AXIS_DIRECT_ENABLE "1"
> dict set params APP_AXIS_SYNC_ENABLE "1"
> dict set params APP_AXIS_IF_ENABLE "1"
> dict set params APP_STAT_ENABLE "1"
> ```
> 
> The application module is responsible for validating the settings (i.e. it should check APP_ID and the various _ENABLE parameters, and fail the build if the configuration is not acceptable). For example, the DMA benchmark application really only requires APP_DMA_ENABLE, all of the others can be turned off as all of those interfaces are either tied off or looped through as appropriate.
> 
> And then the makefile needs to have
> 
> ```
> SYN_FILES += app/dma_bench/rtl/mqnic_app_block_dma_bench.v
> ```
> 
> This one is simple and doesn't require any additional modules beyond the main application module, but in principle other application modules might require more files.



> Thanks for help Alex! This builds the overall design correctly for AU50. However, the mqnic_app_dma_bench.ko is crashing the kernel. The NIC function also stopped working. Is this expected? Thanks!
> 
> ```
> [28427.900941] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: mqnic_app_dma_bench_probe() called
> [28427.900966] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: Allocated DMA region virt 00000000348324b4, phys 00000000652cff22
> [28427.900972] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: Statistics counters
> [28427.900977] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_rx_tlp_mem_rd: 0
> [28427.900982] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_rx_tlp_mem_wr: 0
> [28427.900986] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_rx_tlp_io: 0
> [28427.900990] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_rx_tlp_cfg: 0
> [28427.900995] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_rx_tlp_msg: 0
> [28427.900998] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_rx_tlp_cpl: 0
> [28427.901002] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_rx_tlp_cpl_ur: 0
> [28427.901006] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_rx_tlp_cpl_ca: 0
> [28427.901010] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_rx_tlp_atomic: 0
> [28427.901014] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_rx_tlp_ep: 0
> [28427.901018] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_rx_tlp_hdr_dw: 0
> [28427.901021] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_rx_tlp_req_dw: 0
> [28427.901025] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_rx_tlp_payload_dw: 0
> [28427.901029] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_rx_tlp_cpl_dw: 0
> [28427.901033] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_tx_tlp_mem_rd: 0
> [28427.901037] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_tx_tlp_mem_wr: 0
> [28427.901041] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_tx_tlp_io: 0
> [28427.901044] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_tx_tlp_cfg: 0
> [28427.901048] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_tx_tlp_msg: 0
> [28427.901052] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_tx_tlp_cpl: 0
> [28427.901056] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_tx_tlp_cpl_ur: 0
> [28427.901059] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_tx_tlp_cpl_ca: 0
> [28427.901063] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_tx_tlp_atomic: 0
> [28427.901067] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_tx_tlp_ep: 0
> [28427.901071] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_tx_tlp_hdr_dw: 0
> [28427.901074] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_tx_tlp_req_dw: 0
> [28427.901078] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_tx_tlp_payload_dw: 0
> [28427.901082] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: pcie_tx_tlp_cpl_dw: 0
> [28427.901086] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_rd_op_count: 0
> [28427.901089] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_rd_op_bytes: 0
> [28427.901093] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_rd_op_latency: 0
> [28427.901097] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_rd_op_error: 0
> [28427.901101] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_rd_req_count: 0
> [28427.901104] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_rd_req_latency: 0
> [28427.901108] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_rd_req_timeout: 0
> [28427.901112] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_rd_op_table_full: 0
> [28427.901116] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_rd_no_tags: 0
> [28427.901120] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_rd_tx_no_credit: 0
> [28427.901124] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_rd_tx_limit: 0
> [28427.901128] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_rd_tx_stall: 0
> [28427.901132] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_wr_op_count: 0
> [28427.901136] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_wr_op_bytes: 0
> [28427.901140] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_wr_op_latency: 0
> [28427.901144] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_wr_op_error: 0
> [28427.901147] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_wr_req_count: 0
> [28427.901151] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_wr_req_latency: 0
> [28427.901155] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_wr_op_table_full: 0
> [28427.901159] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_wr_tx_no_credit: 0
> [28427.901163] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_wr_tx_limit: 0
> [28427.901167] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: dma_wr_tx_stall: 0
> [28427.901169] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: Write test data
> [28427.901172] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: Read test data
> [28427.901174] 00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f  ................
> [28427.901177] 10 11 12 13 14 15 16 17 18 19 1a 1b 1c 1d 1e 1f  ................
> [28427.901179] 20 21 22 23 24 25 26 27 28 29 2a 2b 2c 2d 2e 2f   !"#$%&'()*+,-./
> [28427.901180] 30 31 32 33 34 35 36 37 38 39 3a 3b 3c 3d 3e 3f  0123456789:;<=>?
> [28427.901182] 40 41 42 43 44 45 46 47 48 49 4a 4b 4c 4d 4e 4f  @ABCDEFGHIJKLMNO
> [28427.901184] 50 51 52 53 54 55 56 57 58 59 5a 5b 5c 5d 5e 5f  PQRSTUVWXYZ[\]^_
> [28427.901185] 60 61 62 63 64 65 66 67 68 69 6a 6b 6c 6d 6e 6f  `abcdefghijklmno
> [28427.901187] 70 71 72 73 74 75 76 77 78 79 7a 7b 7c 7d 7e 7f  pqrstuvwxyz{|}~.
> [28427.901189] 80 81 82 83 84 85 86 87 88 89 8a 8b 8c 8d 8e 8f  ................
> [28427.901191] 90 91 92 93 94 95 96 97 98 99 9a 9b 9c 9d 9e 9f  ................
> [28427.901192] a0 a1 a2 a3 a4 a5 a6 a7 a8 a9 aa ab ac ad ae af  ................
> [28427.901194] b0 b1 b2 b3 b4 b5 b6 b7 b8 b9 ba bb bc bd be bf  ................
> [28427.901196] c0 c1 c2 c3 c4 c5 c6 c7 c8 c9 ca cb cc cd ce cf  ................
> [28427.901197] d0 d1 d2 d3 d4 d5 d6 d7 d8 d9 da db dc dd de df  ................
> [28427.901199] e0 e1 e2 e3 e4 e5 e6 e7 e8 e9 ea eb ec ed ee ef  ................
> [28427.901201] f0 f1 f2 f3 f4 f5 f6 f7 f8 f9 fa fb fc fd fe ff  ................
> [28427.901203] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: Start copy to card
> [28427.901211] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: Start copy to host
> [28427.901217] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: read test data
> [28427.901219] 00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f  ................
> [28427.901221] 10 11 12 13 14 15 16 17 18 19 1a 1b 1c 1d 1e 1f  ................
> [28427.901223] 20 21 22 23 24 25 26 27 28 29 2a 2b 2c 2d 2e 2f   !"#$%&'()*+,-./
> [28427.901224] 30 31 32 33 34 35 36 37 38 39 3a 3b 3c 3d 3e 3f  0123456789:;<=>?
> [28427.901226] 40 41 42 43 44 45 46 47 48 49 4a 4b 4c 4d 4e 4f  @ABCDEFGHIJKLMNO
> [28427.901228] 50 51 52 53 54 55 56 57 58 59 5a 5b 5c 5d 5e 5f  PQRSTUVWXYZ[\]^_
> [28427.901230] 60 61 62 63 64 65 66 67 68 69 6a 6b 6c 6d 6e 6f  `abcdefghijklmno
> [28427.901231] 70 71 72 73 74 75 76 77 78 79 7a 7b 7c 7d 7e 7f  pqrstuvwxyz{|}~.
> [28427.901233] 80 81 82 83 84 85 86 87 88 89 8a 8b 8c 8d 8e 8f  ................
> [28427.901235] 90 91 92 93 94 95 96 97 98 99 9a 9b 9c 9d 9e 9f  ................
> [28427.901236] a0 a1 a2 a3 a4 a5 a6 a7 a8 a9 aa ab ac ad ae af  ................
> [28427.901238] b0 b1 b2 b3 b4 b5 b6 b7 b8 b9 ba bb bc bd be bf  ................
> [28427.901240] c0 c1 c2 c3 c4 c5 c6 c7 c8 c9 ca cb cc cd ce cf  ................
> [28427.901242] d0 d1 d2 d3 d4 d5 d6 d7 d8 d9 da db dc dd de df  ................
> [28427.901244] e0 e1 e2 e3 e4 e5 e6 e7 e8 e9 ea eb ec ed ee ef  ................
> [28427.901246] f0 f1 f2 f3 f4 f5 f6 f7 f8 f9 fa fb fc fd fe ff  ................
> [28427.901248] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: test data matches
> [28427.901250] mqnic_app_dma_bench.mqnic_app_dma_bench mqnic.app_12348001.0: perform block reads (dma_alloc_coherent)
> [28427.903597] divide error: 0000 [#1] PREEMPT SMP NOPTI
> [28427.903635] CPU: 12 PID: 62474 Comm: insmod Tainted: G        W IOE     5.19.0-rc1+ #7
> [28427.903671] Hardware name: Dell Inc. PowerEdge R740xd/0YNX56, BIOS 2.15.1 06/15/2022
> [28427.903701] RIP: 0010:dma_block_read_bench.constprop.0+0x303/0x336 [mqnic_app_dma_bench]
> [28427.903738] Code: 69 c5 00 b4 c4 04 4e 8d 0c a5 00 00 00 00 4c 2b 45 c0 48 c7 c6 68 46 c9 c0 49 f7 f1 31 d2 50 4a 8d 04 85 00 00 00 00 4d 89 f0 <49> f7 f7 31 d2 50 48 8b 45 d0 41 57 48 c1 e0 02 48 f7 75 b8 ba 10
> [28427.903803] RSP: 0018:ffffa3b025b1f960 EFLAGS: 00010246
> [28427.903830] RAX: 0000000000000000 RBX: ffff8ed39e401228 RCX: 0000000000000001
> [28427.903860] RDX: 0000000000000000 RSI: ffffffffc0c94668 RDI: ffff8ed39ddef800
> [28427.903890] RBP: ffffa3b025b1f9b0 R08: 0000000000000001 R09: 0000000000232208
> [28427.903917] R10: ffffffffffffffff R11: 0000000000000067 R12: 000000000008c882
> [28427.903945] R13: 0000000000000001 R14: 0000000000000001 R15: 0000000000000000
> [28427.903974] FS:  00007f6c08151000(0000) GS:ffff8ed2c0f80000(0000) knlGS:0000000000000000
> [28427.904006] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
> [28427.904031] CR2: 0000564e70115c50 CR3: 00000001143a6001 CR4: 00000000007706e0
> [28427.904059] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
> [28427.904087] DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
> [28427.904116] PKRU: 55555554
> [28427.904130] Call Trace:
> [28427.904145]  <TASK>
> [28427.904161]  mqnic_app_dma_bench_probe+0x44b/0x693 [mqnic_app_dma_bench]
> [28427.904194]  ? acpi_dev_pm_attach+0x25/0x110
> [28427.904223]  ? dma_block_read_bench.constprop.0+0x336/0x336 [mqnic_app_dma_bench]
> [28427.904256]  auxiliary_bus_probe+0x47/0x80
> [28427.904281]  ? driver_sysfs_add+0x77/0xe0
> [28427.904305]  really_probe+0x1d0/0x3b0
> [28427.904328]  __driver_probe_device+0x116/0x190
> [28427.904352]  driver_probe_device+0x23/0xc0
> [28427.904375]  __driver_attach+0xbd/0x1e0
> [28427.904398]  ? __device_attach_driver+0x110/0x110
> [28427.904423]  bus_for_each_dev+0x7c/0xc0
> [28427.904446]  driver_attach+0x1e/0x30
> [28427.904468]  bus_add_driver+0x171/0x210
> [28427.904490]  driver_register+0x95/0xf0
> [28427.904513]  __auxiliary_driver_register+0x6f/0xd0
> [28427.904537]  ? 0xffffffffc0cbd000
> [28427.904556]  mqnic_app_dma_bench_init+0x23/0x1000 [mqnic_app_dma_bench]
> [28427.904587]  do_one_initcall+0x46/0x210
> [28427.904613]  ? kmem_cache_alloc_trace+0x1a6/0x320
> [28427.904642]  do_init_module+0x52/0x210
> [28427.904666]  load_module+0x2012/0x2420
> [28427.904693]  __do_sys_finit_module+0xc5/0x130
> [28427.904716]  ? __do_sys_finit_module+0xc5/0x130
> [28427.904744]  __x64_sys_finit_module+0x18/0x20
> [28427.904767]  do_syscall_64+0x59/0x80
> [28427.905521]  ? ksys_mmap_pgoff+0x10d/0x250
> [28427.906201]  ? exit_to_user_mode_prepare+0x35/0x170
> [28427.906880]  ? syscall_exit_to_user_mode+0x26/0x50
> [28427.907545]  ? __x64_sys_mmap+0x33/0x40
> [28427.908195]  ? do_syscall_64+0x69/0x80
> [28427.908832]  entry_SYSCALL_64_after_hwframe+0x46/0xb0
> [28427.909465] RIP: 0033:0x7f6c0791ea3d
> [28427.910073] Code: 5b 41 5c c3 66 0f 1f 84 00 00 00 00 00 f3 0f 1e fa 48 89 f8 48 89 f7 48 89 d6 48 89 ca 4d 89 c2 4d 89 c8 4c 8b 4c 24 08 0f 05 <48> 3d 01 f0 ff ff 73 01 c3 48 8b 0d c3 a3 0f 00 f7 d8 64 89 01 48
> [28427.911317] RSP: 002b:00007fffcbaf2d18 EFLAGS: 00000246 ORIG_RAX: 0000000000000139
> [28427.911885] RAX: ffffffffffffffda RBX: 0000564e715a27e0 RCX: 00007f6c0791ea3d
> [28427.912147] RDX: 0000000000000000 RSI: 0000564e7011ecd2 RDI: 0000000000000003
> [28427.912412] RBP: 0000000000000000 R08: 0000000000000000 R09: 0000000000000000
> [28427.912676] R10: 0000000000000003 R11: 0000000000000246 R12: 0000564e7011ecd2
> [28427.912937] R13: 0000564e715a2760 R14: 0000564e7011d888 R15: 0000564e715a2900
> [28427.913201]  </TASK>
> [28427.913453] Modules linked in: mqnic_app_dma_bench(OE+) tls mqnic(OE) nfsv3 nfs_acl rpcsec_gss_krb5 nfsv4 nfs lockd grace fscache netfs binfmt_misc nvme_fabrics intel_rapl_msr intel_rapl_common intel_uncore_frequency intel_uncore_frequency_common isst_if_common skx_edac nfit x86_pkg_temp_thermal intel_powerclamp coretemp kvm_intel ipmi_ssif kvm cmdlinepart spi_nor rapl dell_smbios mei_me dcdbas intel_cstate joydev dell_wmi_descriptor wmi_bmof input_leds mei mtd intel_pch_thermal acpi_ipmi ipmi_si acpi_power_meter mac_hid auth_rpcgss sch_fq_codel sunrpc ipmi_devintf dm_multipath ipmi_msghandler scsi_dh_rdac scsi_dh_emc scsi_dh_alua ramoops msr pstore_blk reed_solomon pstore_zone efi_pstore ip_tables x_tables autofs4 btrfs blake2b_generic zstd_compress raid10 raid456 async_raid6_recov async_memcpy async_pq async_xor async_tx xor raid6_pq libcrc32c raid1 raid0 multipath linear mlx4_ib ib_uverbs ib_core mlx4_en hid_generic usbhid hid mgag200 drm_shmem_helper drm_kms_helper syscopyarea
> [28427.913510]  sysfillrect crct10dif_pclmul crc32_pclmul ghash_clmulni_intel sysimgblt nvme fb_sys_fops ixgbe aesni_intel igb xfrm_algo crypto_simd spi_intel_pci bnxt_en mlx4_core i2c_i801 drm cryptd nvme_core megaraid_sas ahci xhci_pci dca spi_intel lpc_ich i2c_smbus libahci xhci_pci_renesas wmi i2c_algo_bit mdio
> [28427.916972] ---[ end trace 0000000000000000 ]---
> [28427.954067] RIP: 0010:dma_block_read_bench.constprop.0+0x303/0x336 [mqnic_app_dma_bench]
> [28427.954456] Code: 69 c5 00 b4 c4 04 4e 8d 0c a5 00 00 00 00 4c 2b 45 c0 48 c7 c6 68 46 c9 c0 49 f7 f1 31 d2 50 4a 8d 04 85 00 00 00 00 4d 89 f0 <49> f7 f7 31 d2 50 48 8b 45 d0 41 57 48 c1 e0 02 48 f7 75 b8 ba 10
> [28427.954971] RSP: 0018:ffffa3b025b1f960 EFLAGS: 00010246
> [28427.955236] RAX: 0000000000000000 RBX: ffff8ed39e401228 RCX: 0000000000000001
> [28427.955502] RDX: 0000000000000000 RSI: ffffffffc0c94668 RDI: ffff8ed39ddef800
> [28427.955784] RBP: ffffa3b025b1f9b0 R08: 0000000000000001 R09: 0000000000232208
> [28427.956053] R10: ffffffffffffffff R11: 0000000000000067 R12: 000000000008c882
> [28427.956322] R13: 0000000000000001 R14: 0000000000000001 R15: 0000000000000000
> [28427.956593] FS:  00007f6c08151000(0000) GS:ffff8ed2c0f80000(0000) knlGS:0000000000000000
> [28427.956868] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
> [28427.957143] CR2: 0000564e70115c50 CR3: 00000001143a6001 CR4: 00000000007706e0
> [28427.957419] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
> [28427.957695] DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
> [28427.957982] PKRU: 55555554
> ```

Hi Alex,
I have been trying to enable the mqnic_app_dma_bench application for AU50, I had configured the config.tcl file as follows and similarly in the driver code. I has used this repo which you have mentioned in this issue #128
https://github.com/ucsdsysnet/corundum
> # Application block configuration
> dict set params APP_ID "32'h12348101"
> dict set params APP_ENABLE "1"
> dict set params APP_CTRL_ENABLE "1"
> dict set params APP_DMA_ENABLE "1"
> dict set params APP_AXIS_DIRECT_ENABLE "1"
> dict set params APP_AXIS_SYNC_ENABLE "1"
> dict set params APP_AXIS_IF_ENABLE "1"
> dict set params APP_STAT_ENABLE "1"
However I get below error , DMA bench register block not found. Could you let me know what am I missing here. The simulation testcase in this path ucsdsysnet/corundum/fpga/app/dma_bench/tb/mqnic_core_pcie_us works fine with test PASS 
for this application.

<img width="1069" height="945" alt="Image" src="https://github.com/user-attachments/assets/fd549f1f-7ca2-4aca-bcb9-a24b6692687e" />

**Comment by alexforencich** [comment]
Looks like all of the statistics counters read as zero, and that resulted in a divide-by-zero triggering a panic.  Not sure why they would all read as zero though.  The NIC functionality should still work, although that panic could have caused some problems in the kernel.

**Comment by lizajoseph** [comment]
> Looks like all of the statistics counters read as zero, and that resulted in a divide-by-zero triggering a panic. Not sure why they would all read as zero though. The NIC functionality should still work, although that panic could have caused some problems in the kernel.

Hi Alex,
What are the statistics counters that you are referring to here ?

**Comment by lizajoseph** [comment]
> Looks like all of the statistics counters read as zero, and that resulted in a divide-by-zero triggering a panic. Not sure why they would all read as zero though. The NIC functionality should still work, although that panic could have caused some problems in the kernel.

Hi Alex,

I was able to get this issue resolved. The issue was that I had added this custom application logic verilog file /corundum/fpga/app/template/rtl/mqnic_app_block.v instead of /corundum/fpga/app/dma_bench/rtl/mqnic_app_block_dma_bench.v to run and test the DMA benchmarking application. Once this mqnic_app_block_dma_bench.v file was added in the Makefile the custom APP block got correctly instanced in vivado and with loading mqnic.ko and mqnic_app_dma_bench.ko the application works as expected.

Thanks
